File: services/payment/src/services/PaymentService.ts
import { injectable, inject } from "inversify";
import { IPaymentRepository } from "../repositories/PaymentRepository";
import { Payment } from "../entities/Payment.entity";
import { ILogger, TYPES } from "@observability/core";
import {
  ProcessPaymentDto,
  PaymentStatus,
  PaymentError,
  PaymentSimulator,
} from "../utils/payment.utils";
import { randomUUID } from "crypto";

export interface IPaymentService {
  processPayment(paymentDto: ProcessPaymentDto, traceId?: string): Promise<Payment>;
  getPaymentById(id: string, traceId?: string): Promise<Payment | null>;
  getPaymentByOrderId(orderId: string, traceId?: string): Promise<Payment | null>;
  getAllPayments(traceId?: string): Promise<Payment[]>;
}

@injectable()
export class PaymentService implements IPaymentService {
  constructor(
    @inject("PaymentRepository") private paymentRepo: IPaymentRepository,
    @inject(TYPES.Logger) private logger: ILogger
  ) {}

  async processPayment(paymentDto: ProcessPaymentDto, traceId?: string): Promise<Payment> {
    const startTime = Date.now();
    const transactionId = `TXN-${randomUUID()}`;
    try {
      this.logger.info("Processing payment", { traceId, transactionId, orderId: paymentDto.orderId, amount: paymentDto.amount });
      const payment = await this.paymentRepo.create({
        orderId: paymentDto.orderId,
        transactionId,
        amount: paymentDto.amount,
        currency: paymentDto.currency,
        method: paymentDto.method,
        customerDetails: paymentDto.customerDetails,
        status: PaymentStatus.PROCESSING,
        processingTimeMs: 0,
      });
      const delay = PaymentSimulator.simulateProcessingDelay();
      await this.sleep(delay);
      if (PaymentSimulator.shouldSimulateFailure()) {
        return await this.handlePaymentFailure(payment.id, startTime, traceId);
      }
      const processingTime = Date.now() - startTime;
      const updatedPayment = await this.paymentRepo.updateStatus(payment.id, PaymentStatus.SUCCESS);
      await this.updateProcessingTime(payment.id, processingTime);
      this.logger.info("Payment successful", { traceId, processingTime, transactionId, orderId: paymentDto.orderId });
      return updatedPayment!;
    } catch (error) {
      this.logger.error("Payment processing error", error as Error, { traceId, transactionId, error: (error as Error).message, stack: (error as Error).stack });
      throw error;
    }
  }

  async getPaymentById(id: string, traceId?: string): Promise<Payment | null> {
    this.logger.info("Get payment by id", { traceId, id });
    return await this.paymentRepo.findById(id);
  }

  async getPaymentByOrderId(orderId: string, traceId?: string): Promise<Payment | null> {
    this.logger.info("Get payment by orderId", { traceId, orderId });
    return await this.paymentRepo.findByOrderId(orderId);
  }

  async getAllPayments(traceId?: string): Promise<Payment[]> {
    this.logger.info("Get all payments", { traceId });
    return await this.paymentRepo.findAll();
  }

  private async handlePaymentFailure(paymentId: string, startTime: number, traceId?: string): Promise<Payment> {
    const failureReason = PaymentSimulator.getRandomFailureReason();
    const processingTime = Date.now() - startTime;
    this.logger.warn("Payment failed", { traceId, paymentId, reason: failureReason });
    const payment = await this.paymentRepo.updateStatus(paymentId, PaymentStatus.FAILED, failureReason);
    await this.updateProcessingTime(paymentId, processingTime);
    throw new PaymentError(failureReason, "PAYMENT_FAILED");
  }

  private async updateProcessingTime(paymentId: string, timeMs: number): Promise<void> {
    const repo = this.paymentRepo as any;
    const payment = await this.paymentRepo.findById(paymentId);
    if (payment) {
      payment.processingTimeMs = timeMs;
      await repo.getRepository().save(payment);
    }
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: services-java/payment/src/main/java/org/observability/service/PaymentService.java
package org.observability.service;

import org.observability.dto.ProcessPaymentDto;
import org.observability.entity.Payment;
import org.observability.enums.PaymentStatus;
import org.observability.exception.PaymentException;
import org.observability.repository.PaymentRepository;
import org.observability.util.PaymentSimulator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
public class PaymentService {

    private static final Logger logger = LoggerFactory.getLogger(PaymentService.class);

    private final PaymentRepository paymentRepository;

    @Autowired
    public PaymentService(PaymentRepository paymentRepository) {
        this.paymentRepository = paymentRepository;
    }

    @Transactional
    public Payment processPayment(ProcessPaymentDto paymentDto, String traceId) {
        long startTime = System.currentTimeMillis();
        String transactionId = "TXN-" + UUID.randomUUID().toString();

        try {
            logger.info("Processing payment - TraceId: {}, TransactionId: {}, OrderId: {}, Amount: {}",
                    traceId, transactionId, paymentDto.getOrderId(), paymentDto.getAmount());

            // Create payment (initial state)
            Payment payment = new Payment();
            payment.setOrderId(UUID.fromString(paymentDto.getOrderId()));
            payment.setTransactionId(transactionId);
            payment.setAmount(paymentDto.getAmount());
            payment.setCurrency(paymentDto.getCurrency());
            payment.setMethod(paymentDto.getMethod());
            payment.setStatus(PaymentStatus.PROCESSING);
            payment.setProcessingTimeMs(0);

            // Convert customer details to Map
            Map<String, String> customerDetails = new HashMap<>();
            customerDetails.put("name", paymentDto.getCustomerDetails().getName());
            customerDetails.put("email", paymentDto.getCustomerDetails().getEmail());
            payment.setCustomerDetails(customerDetails);

            Payment savedPayment = paymentRepository.save(payment);

            // Simulate processing delay
            int delay = PaymentSimulator.simulateProcessingDelay();
            sleep(delay);

            // Simulate failure (40% chance)
            if (PaymentSimulator.shouldSimulateFailure()) {
                return handlePaymentFailure(savedPayment.getId(), startTime, traceId);
            }

            // Mark as SUCCESS
            long processingTime = System.currentTimeMillis() - startTime;
            savedPayment.setStatus(PaymentStatus.SUCCESS);
            savedPayment.setProcessingTimeMs((int) processingTime);
            Payment updatedPayment = paymentRepository.save(savedPayment);

            logger.info("Payment successful - TraceId: {}, ProcessingTime: {}ms, TransactionId: {}, OrderId: {}",
                    traceId, processingTime, transactionId, paymentDto.getOrderId());

            return updatedPayment;

        } catch (PaymentException e) {
            throw e;
        } catch (Exception e) {
            logger.error("Payment processing error - TraceId: {}, TransactionId: {}, Error: {}",
                    traceId, transactionId, e.getMessage(), e);
            throw new PaymentException("Payment processing failed: " + e.getMessage(), "PAYMENT_ERROR");
        }
    }

    public Optional<Payment> getPaymentById(UUID id, String traceId) {
        logger.info("Get payment by ID - TraceId: {}, PaymentId: {}", traceId, id);
        return paymentRepository.findById(id);
    }

    public Optional<Payment> getPaymentByOrderId(UUID orderId, String traceId) {
        logger.info("Get payment by OrderId - TraceId: {}, OrderId: {}", traceId, orderId);
        return paymentRepository.findByOrderId(orderId);
    }

    public List<Payment> getAllPayments(String traceId) {
        logger.info("Get all payments - TraceId: {}", traceId);
        return paymentRepository.findAllByOrderByCreatedAtDesc();
    }

    private Payment handlePaymentFailure(UUID paymentId, long startTime, String traceId) {
        String failureReason = PaymentSimulator.getRandomFailureReason();
        long processingTime = System.currentTimeMillis() - startTime;

        logger.warn("Payment failed - TraceId: {}, PaymentId: {}, Reason: {}", traceId, paymentId, failureReason);

        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new PaymentException("Payment not found", "PAYMENT_NOT_FOUND"));

        payment.setStatus(PaymentStatus.FAILED);
        payment.setFailureReason(failureReason);
        payment.setProcessingTimeMs((int) processingTime);
        paymentRepository.save(payment);

        throw new PaymentException(failureReason, "PAYMENT_FAILED");
    }

    private void sleep(int ms) {
        try {
            Thread.sleep(ms);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new PaymentException("Payment processing interrupted", "PROCESSING_INTERRUPTED");
        }
    }
}
